src/views/GameLobbyScreen/GameLobbyScreen.js

```diff
@@ -0,0 +1,193 @@
+import React, { Component, PropTypes } from 'react';
+import { bindActionCreators, compose } from 'redux';
+import { connect } from 'react-redux';
+import { withRouter } from 'react-router';
+
+import { withSocket } from 'hoc';
+import { GameLobby } from 'components';
+import { setReady, toggleReady } from 'redux/modules/ready';
+import { startGame, updateHasOpponent, resetCurrentGame } from 'redux/modules/currentGame';
+import { setOpponentName } from 'redux/modules/name';
+import { openFriendInviteModal, closeFriendInviteModal } from 'redux/modules/friendInviteModal';
+
+export class GameLobbyScreen extends Component {
+  static propTypes = {
+    player: PropTypes.shape({
+      name: PropTypes.string.isRequired,
+      ready: PropTypes.bool.isRequired,
+    }).isRequired,
+    opponent: PropTypes.shape({
+      name: PropTypes.string.isRequired,
+      ready: PropTypes.bool.isRequired,
+    }).isRequired,
+    gameId: PropTypes.string.isRequired,
+    hasOpponent: PropTypes.bool.isRequired,
+    started: PropTypes.bool.isRequired,
+    friendInviteModal: PropTypes.shape({
+      isOpen: PropTypes.bool.isRequired,
+    }).isRequired,
+    actions: PropTypes.shape({
+      setReady: PropTypes.func.isRequired,
+      toggleReady: PropTypes.func.isRequired,
+      startGame: PropTypes.func.isRequired,
+      updateHasOpponent: PropTypes.func.isRequired,
+      setOpponentName: PropTypes.func.isRequired,
+      resetCurrentGame: PropTypes.func.isRequired,
+    }),
+    playerCardActions: PropTypes.shape({
+      openFriendInviteModal: PropTypes.func.isRequired,
+      closeFriendInviteModal: PropTypes.func.isRequired,
+    }),
+    router: PropTypes.shape({
+      push: PropTypes.func.isRequired,
+      createHref: PropTypes.func.isRequired,
+    }).isRequired,
+    socket: PropTypes.shape({
+      id: PropTypes.string.isRequired,
+      emit: PropTypes.func.isRequired,
+      on: PropTypes.func.isRequired,
+    }).isRequired,
+  }
+
+  state = {
+    canCountdown: true,
+  }
+
+  componentDidMount = () => {
+    this.joinGame(this.props);
+    this.notifyOnPlayerJoined(this.props);
+  }
+
+  componentWillReceiveProps = (nextProps) => {
+    if (nextProps.gameId !== this.props.gameId) {
+      this.joinGame(nextProps);
+    }
+
+    if (nextProps.player.ready && nextProps.opponent.ready) {
+      this.startCountdown();
+    } else {
+      this.stopCountdown();
+    }
+  }
+
+  componentWillUnmount = () => {
+    if (!this.props.started) {
+      this.leaveGame(this.props);
+    }
+
+    this.removeOnPlayerJoinedListener(this.props);
+  }
+
+  startCountdown = () => this.setState(() => ({ canCountdown: true }));
+  stopCountdown = () => this.setState(() => ({ canCountdown: false }));
+
+  handleCountdownFinish = () => {
+    this.startGame(this.props);
+    this.goInGame(this.props);
+  }
+
+  notifyOnPlayerJoined = (props) => {
+    props.socket.on('playerJoined', ({ socketId, name, playerCount }) => {
```

Comment:
We should definitely abstract away the `socket` instance here, instead of supplying the actual instance here. We should only expose methods that allow for channels that we support. No other channels. This makes it easier to use and doesn't require the consumers to know what channels to talk to. They only check the API and that's enough.

Reply:
We could also move these socket `on` calls to a separate React component, to manage the event callbacks within the React lifecycle.

**Example:**

```js
class ListenPlayerJoined extends React.Component {
  constructor(props, ...args) {
    super(props, ...args);

    this.callback = this.props.callback;
  }

  componentDidMount() {
    this.props.socket.on('playerJoined', this.callback);
  }

  componentWillUnmount() {
    this.props.socket.removeListener('playerJoined', this.callback);
  }

  render() {
    return null;
  }
}
```

```js
// consumer
<SocketManager>
  <ListenPlayerJoined callback={} />
</SocketManager>
```